website/blog/2023-01-20-typescript-first.md

```diff
@@ -0,0 +1,101 @@
+- [TypeScript declarations shipped with React Native](/blog/2023/01/20/typescript-first#declarations-shipped-with-react-native)
+- [React Native documentation is TypeScript First](/blog/2023/01/20/typescript-first#documentation-is-typescript-first)
+
+In this post we’ll cover what these changes mean for you as a TypeScript or Flow user.
+
+<!--truncate-->
+
+## New App Template is TypeScript By Default
+
+Starting with 0.71, when you create a new React Native app via the React Native CLI you'll get a TypeScript app by default!
+
+```bash
+$ npx react-native init My71App --version 0.71.0
+```
+
+![Screenshot of an iPhone simulator running a new app generated by React Native CLI. Alongside the simulator is a screenshot of Visual Studio Code editor opened to "App.tsx" to illustrate it is running a TypeScript file.](/blog/assets/typescript-first-new-app.png)
+
+The starting point of a newly generated app will be `App.tsx` instead of `App.js` – fully TypeScript typed. The new project is already set up with a `tsconfig.json` so out of the box your IDE will help you write typed code right away!
+
+## Declarations Shipped with React Native
+
+0.71 is the first release with built-in Typescript (TS) declarations.
+
+Prior, TypeScript declarations for React Native were provided by [`@types/react-native`](https://www.npmjs.com/package/@types/react-native) hosted in the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repository. The decision to co-locate TypeScript types with React Native source was for improved correctness and maintenance.
+
+`@types/react-native` only provides types for stable releases. This means if you ever wanted to develop with a pre-release version of React Native in TypeScript, you’d have to use types from an older release which may be inaccurate. Releasing `@types/react-native` is also error-prone. The releases lag React Native releases, and the process involves manually inspecting for type changes made to React Native’s public API and updating the TS declaration to match.
+
+With TS types co-located with React Native source, there is more visibility and ownership of TS types. Our team is actively working on tools to maintain alignment between Flow and TS.
+
+This change also removes a dependency for React Native users to manage. When upgrading to 0.71 or above, you can remove `@types/react-native` as a dependency. [Refer to the new app template on how to set up TypeScript support.](https://github.com/facebook/react-native/blob/main/template/tsconfig.json)
+
+We plan to deprecate `@types/react-native` for versions 0.73 and onward. Concretely this means:
+
+- `@types/react-native` tracking React Native versions 0.71 and 0.72 will be released. They will be identical to the types in React Native on the relevant release branches.
+- For React Native 0.73 and onward, TS types will only be available from React Native.
```

Comment:
OT from the content, I knew we were planning to provide @types/react-native for 0.71, but didn't realize we were also for 0.72. Are we wanting to give more time for migration/figuring out the story with DT deprecation?

Reply:
That was my intention but I don't feel strongly. I don't think the cost of maintaining 0.72 feels that high. It's mostly about making sure relevant changes sent to DefinitelyTyped are shared in React Native.

Reply:
Makes sense to me. I will be interested to see how many folks latch onto the 0.71 DefinitelyTyped typings. We will need to generate signal to retrain folks, and this blog post is a great start.

From the docs I read on how TS does module resolution I think the built-in types will win against the DefinitelyTyped ones if both are present, and a module tries to resolve types for "react-native". So if someone submits a fix to DefinitelyTyped but not RN it might just no-op (given RN is also present).